From f43489b08c07e4aa6090a79468e44eb6fd95313e Mon Sep 17 00:00:00 2001
From: Wang Wensheng <wangwensheng4@huawei.com>
Date: Thu, 14 Jul 2022 11:57:51 +0000
Subject: [PATCH 1/5] ascend/arm64: Add ascend_enable_all kernel paramater

This kernel paramater is used for ascend scene and would open all the
options needed at once.

Signed-off-by: Wang Wensheng <wangwensheng4@huawei.com>
---
 arch/arm64/kernel/cpufeature.c |  2 +-
 arch/arm64/mm/init.c           | 38 ++++++++++++++++++++++++++++++++++
 mm/hugetlb.c                   |  2 +-
 mm/memcontrol.c                |  2 +-
 4 files changed, 41 insertions(+), 3 deletions(-)

diff --git a/arch/arm64/kernel/cpufeature.c b/arch/arm64/kernel/cpufeature.c
index f5ce1e3a5..159481996 100644
--- a/arch/arm64/kernel/cpufeature.c
+++ b/arch/arm64/kernel/cpufeature.c
@@ -1729,7 +1729,7 @@ static void cpu_enable_e0pd(struct arm64_cpu_capabilities const *cap)
 #endif /* CONFIG_ARM64_E0PD */
 
 #ifdef CONFIG_ARM64_PSEUDO_NMI
-static bool enable_pseudo_nmi;
+bool enable_pseudo_nmi;
 
 static int __init early_enable_pseudo_nmi(char *p)
 {
diff --git a/arch/arm64/mm/init.c b/arch/arm64/mm/init.c
index 7119dd619..21588b481 100644
--- a/arch/arm64/mm/init.c
+++ b/arch/arm64/mm/init.c
@@ -31,6 +31,8 @@
 #include <linux/hugetlb.h>
 #include <linux/acpi_iort.h>
 #include <linux/pin_mem.h>
+#include <linux/suspend.h>
+#include <linux/nmi.h>
 
 #include <asm/boot.h>
 #include <asm/fixmap.h>
@@ -817,3 +819,39 @@ void dump_mem_limit(void)
 		pr_emerg("Memory Limit: none\n");
 	}
 }
+
+void ascend_enable_all_features(void)
+{
+	if (IS_ENABLED(CONFIG_ASCEND_DVPP_MMAP))
+		enable_mmap_dvpp = 1;
+
+#ifdef CONFIG_ASCEND_CHARGE_MIGRATE_HUGEPAGES
+	extern int enable_charge_mighp;
+	enable_charge_mighp = 1;
+#endif
+
+	if (IS_ENABLED(CONFIG_SUSPEND))
+		mem_sleep_current = PM_SUSPEND_ON;
+
+	if (IS_ENABLED(CONFIG_MEMCG_KMEM)) {
+		extern bool cgroup_memory_nokmem;
+		cgroup_memory_nokmem = false;
+	}
+
+#ifdef CONFIG_ARM64_PSEUDO_NMI
+	extern bool enable_pseudo_nmi;
+	enable_pseudo_nmi = true;
+#endif
+
+#ifdef CONFIG_CORELOCKUP_DETECTOR
+	enable_corelockup_detector = true;
+#endif
+}
+
+static int __init ascend_enable_setup(char *__unused)
+{
+	ascend_enable_all_features();
+
+	return 0;
+}
+early_param("ascend_enable_all", ascend_enable_setup);
diff --git a/mm/hugetlb.c b/mm/hugetlb.c
index a8c815386..6cb480539 100644
--- a/mm/hugetlb.c
+++ b/mm/hugetlb.c
@@ -6206,7 +6206,7 @@ void __init hugetlb_cma_check(void)
 #endif /* CONFIG_CMA */
 
 #ifdef CONFIG_ASCEND_FEATURES
-static int enable_charge_mighp __read_mostly;
+int enable_charge_mighp __read_mostly;
 
 const struct hstate *hugetlb_get_hstate(void)
 {
diff --git a/mm/memcontrol.c b/mm/memcontrol.c
index 3bc216a14..fda321322 100644
--- a/mm/memcontrol.c
+++ b/mm/memcontrol.c
@@ -84,7 +84,7 @@ DEFINE_PER_CPU(struct mem_cgroup *, int_active_memcg);
 static bool cgroup_memory_nosocket;
 
 /* Kernel memory accounting disabled */
-static bool cgroup_memory_nokmem = true;
+bool cgroup_memory_nokmem = true;
 
 /* Whether the swap controller is active */
 #ifdef CONFIG_MEMCG_SWAP
-- 
2.17.1

